src/libsyntax/ext/expand.rs

```diff
@@ -640,6 +640,9 @@ pub fn expand_item_mac(it: P<ast::Item>, fld: &mut MacroExpander)
             // need to be marked. Not that it could be marked anyway.
             // create issue to recommend refactoring here?
             fld.cx.syntax_env.insert(intern(name[]), ext);
+            // Keep track of this macro definition both within the crate
+            // context and (if applicable) in its exports.
+            fld.cx.macros.push(it.clone());
```

Comment:
Unfortunately [`rustc::sess::Session#add_lint`](http://doc.rust-lang.org/rustc/session/struct.Session.html#method.add_lint) doesn't seem to be available at the original [`add_new_extension`](https://github.com/rust-lang/rust/blob/856b90c400b0a3f5e8018255c7e33d3c1ae84a10/src/libsyntax/ext/tt/macro_rules.rs#L218-L222) check point in libsyntax, and I can't find any way to get to it via the ExtCtxt.

The following idea definitely isn't the most elegant approach: instead of having the whole `.macros` property and checking it with a LintPass, what if the `add_new_extension` function pushed shadowings it found onto a member of `ext::ExtCtxt`, which then got copied onto the `ast::Crate`, and then those got reported by a LintPass later on. That way the checks would be happening in the right place (during syntax expansion), but reporting would be deferred until the proper linting infrastructure was available.